pageserver/src/tenant_mgr.rs

```diff
@@ -147,62 +147,68 @@ pub fn attach_local_tenants(
-                                tenant.update_tenant_config(tenant_conf);
-                                tenant.activate(false);
-                            }
-                            Err(e) => {
-                                error!("Failed to read config for tenant {tenant_id}, disabling tenant: {e:?}");
-                                tenant.set_state(TenantState::Broken);
-                            }
-                        };
-                    }
-                }
-                v.insert(Arc::clone(&tenant));
+                tenant.with_upgrade(|tenant| {
+                    let tenant = tenant.expect("tenant must be alive after creation");
+                    match local_timelines {
+                        TenantAttachData::Broken(_) => {
+                            tenant.set_state(TenantState::Broken);
+                        }
+                        TenantAttachData::Ready(_) => {
+                            match Tenant::load_tenant_config(conf, tenant_id) {
+                                Ok(tenant_conf) => {
+                                    tenant.update_tenant_config(tenant_conf);
+                                    tenant.activate(false);
+                                }
+                                Err(e) => {
+                                    error!("Failed to read config for tenant {tenant_id}, disabling tenant: {e:?}");
+                                    tenant.set_state(TenantState::Broken);
+                                }
+                            };
+                        }
+                    };
+                });
+                v.insert(tenant.clone());
                 tenant
             }
         };
         drop(tenants_accessor);
-        match local_timelines {
-            TenantAttachData::Broken(e) => warn!("{}", e),
-            TenantAttachData::Ready(ref timelines) => {
-                info!("Attaching {} timelines for {tenant_id}", timelines.len());
-                debug!("Timelines to attach: {local_timelines:?}");
-                let has_timelines = !timelines.is_empty();
-                let timelines_to_attach = timelines
-                    .iter()
-                    .map(|(&k, v)| (k, v.metadata().to_owned()))
-                    .collect();
-                match tenant.init_attach_timelines(timelines_to_attach) {
-                    Ok(()) => {
-                        info!("successfully loaded local timelines for tenant {tenant_id}");
-                        tenant.activate(has_timelines);
-                    }
-                    Err(e) => {
-                        error!("Failed to attach tenant timelines: {e:?}");
-                        tenant.set_state(TenantState::Broken);
+        tenant.with_upgrade(|tenant| {
+            let tenant = tenant.expect("tenant must be alive after creation");
+            match local_timelines {
+                TenantAttachData::Broken(e) => warn!("{}", e),
```

Comment:
I think it won't affect performance, but since `TenantAccessor` stores state, this state could be changed somehow by operations that are currently inside the closure. This could lead to missed changes in the state by the accessor and break the invariants.

Probably we should forbid creating tenant without TenantAccessor, i.e private it?